### DISCUSSION TITLE
Best way to create custom keybindings

### DISCUSSION THREAD

**Discussion by akhi3030** [regolith-linux/regolith-desktop#1008] [Q&A] [answered]
I am trying to figure out what is the best way to setup custom keybindings.  I tried to look for documentation in various places but did not find anything addressing the errors I am getting so I am trying out here.

## Context

```
cat /etc/regolith/version 
REGOLITH_VERSION=3.1
```

`/etc/regolith/i3/config` contains the following:

```
# Include common partials
include /usr/share/regolith/common/config.d/*

# Include any regolith i3 partials
include /usr/share/regolith/i3/config.d/*

# Include any partials common between i3 and sway sessions
include $HOME/.config/regolith3/common-wm/config.d/*

# Include any user i3 partials
include $HOME/.config/regolith3/i3/config.d/*
```

Based on this I created `$HOME/.config/regolith3/i3/config.d/config` with the following content

```
bindsym $mod+Tab focus right
```

Upon reloading the configuration, I get the following errors:

```
ERROR: Duplicate keybinding in config file:
  state mask 0x40 with keysym Tab, command "focus right"
ERROR: FYI: You are using i3 version 5b24e18
ERROR: Duplicate keybinding in config file:
  state mask 0x40 with keysym Tab, command "focus right"
ERROR: FYI: You are using i3 version 5b24e18
...
```

Based on this, I saw a duplicate setting in `/usr/share/regolith/common/config.d/30_navigation`
```
...
## Navigate // Next Workspace // <> Tab ##
set_from_resource $wm.binding.ws_next wm.binding.ws_next Tab
bindsym $mod+$wm.binding.ws_next workspace next
...
```

Commenting out the above lines makes the errors go away.

## Questions
So my question are:
- Is this the right way to setup custom keybindings?  
- Do I need to find and comment out the original bindings that overlap with the custom ones?  If I do this, will this cause trouble down the line if I am upgrading regolith (my understanding is that `apt` does not remove modified files and that can cause conflicts).

Any general guidance on how to do it cleanly will be much appreciated!
Thank you very much!

**Comment by ajkerrigan** [comment, marked as answer]
The [how to customize a default partial](https://regolith-desktop.com/docs/howtos/customize-i3-configuration/#how-to-customize-a-default-partial) docs suggest...

* Making sure you have a relevant directory for config overrides: `mkdir -p ~/.config/regolith3/common-wm/config.d`
* Copying `/usr/share/regolith/common/config.d/30_navigation` to `~/.config/regolith3/common-wm/config.d/30_navigation`
* Customizing your keybindings - comment/remove the default Tab binding and add yours
* Remove the package that provides the default partial - in this case `sudo apt remove regolith-wm-navigation`

I've done a bit of this myself since a recent upgrade to 3.1, and have found this pattern to work pretty well. I got into a habit of first searching `/usr/share/regolith` for a given keybinding to find where the default was defined, and then using `dpkg -S` to find out which apt package created that file:

```console
❯ grep -r Tab /usr/share/regolith

/usr/share/regolith/common/config.d/30_navigation:## Navigate // Next Workspace // <> Tab ##
/usr/share/regolith/common/config.d/30_navigation:set_from_resource $wm.binding.ws_next wm.binding.ws_next Tab
/usr/share/regolith/common/config.d/30_navigation:## Navigate // Next Workspace on Output // <><Ctrl> Tab ##
/usr/share/regolith/common/config.d/30_navigation:set_from_resource $wm.binding.ws_next_on_output wm.binding.ws_next_on_output Ctrl+Tab
/usr/share/regolith/common/config.d/30_navigation:## Navigate // Previous Workspace // <><Shift> Tab ##
/usr/share/regolith/common/config.d/30_navigation:set_from_resource $wm.binding.ws_prev wm.binding.ws_prev Shift+Tab
/usr/share/regolith/common/config.d/30_navigation:## Navigate // Previous Workspace on Output // <><Ctrl><Shift> Tab ##
/usr/share/regolith/common/config.d/30_navigation:set_from_resource $wm.binding.ws_prev_on_output wm.binding.ws_prev_on_output Ctrl+Shift+Tab
```

```console
❯ dpkg -S 30_navigation

regolith-wm-navigation: /usr/share/regolith/common/config.d/30_navigation
```

**Reply by akhi3030** [reply]
thank you very much for your detailed answer.  Yes, this flow works for me as well.
